### ISSUE TITLE
Question about Bias Quantification

### ISSUE THREAD

**Issue by MarWaltz** [watchernyu/REDQ#9]
Dear Che Wang,

Thanks again for providing the details.

I came back to your approach of bias quantification, and I think I have kind of a misunderstanding. In the paper, Section 3, you define a Q-function Q^{\pi} for a policy \pi. Further, Q^{\phi} is an estimate of it, and the bias is consequently Q^{\phi} - Q^{\pi}. In the experiment, you consider the current behavior policy \pi induced by your current actor, which contains some noise for exploration (I looked it up in your code). This is the policy for generating your MC trajectories, which serve as the ground truth and are used to compute returns by definition.

But is comparing these returns created by the continuous analog of an \epsilon-greedy strategy with the learned Q's of the critic a valid approach? In Q-Learning, or the critic training in this case, the Q directly approximates Q* (Sutton & Barto, 2018, p. 131), regardless of the behavior policy. So I am wondering how the pursued approach can be justified.

Again, I believe that I have some conceptual misunderstanding, and I would be glad if you could help me with this point.

Best,
Martin

**Comment by watchernyu** [comment]
Hi! Thank you for this very good question! Here are my thoughts on this question: 

1. Q learning directly approximates Q*, this is correct, and if we really want to evaluate the Q value of the current policy we have to use a different algorithm (I think it's some form of SARSA). So the Q estimate we have at any point during training is closely related to the current policy, but it only provides a rough estimate of the value of the current policy. If we naively stop getting more data and keep doing Q updates, we eventually run into the extrapolation error problem, a common issue for offline RL. So there should be a better way to more accurately estimate the true Q value of the current policy, I believe this line of research is called off-policy evaluation. In our paper, we did not investigate very deeply into this direction, instead, we mainly followed what was done in previous works (DDQN, TD3). 

2. Does it make sense to compare MC return of current policy with the Q estimate, in terms of the exploration noise: Due to the design of SAC, they perform what is sometimes called "policy smoothing", which means when their Q networks are being updated, they actually use sampled actions, similar to when sample actions for exploration. So I believe it makes sense to make sure this part is consistent, since the Q network is learned with data that have exploration noise, then during our MC rollouts we should also collect data with exploration noise. 

In summary, you are making a very good observation, and indeed the bias estimation technique in our paper, and in previous works, is not that refined. There might be some good opportunities here for future research. I haven't looked at too many off-policy evaluation literature yet, so I'm not an expert in this particular subtle issue and I could be wrong, but I hope the above answer helps a bit, and I'm happy to further discuss if you have other questions :)

**Comment by MarWaltz** [comment]
Thanks for your extended answer; I appreciate that.

However, I am still interested in how one can adequately quantify the estimation bias during the training of Q-Learning. I agree with you that it is easier for SARSA, which is on-policy. We behave according to some policy _pi_ and simultaneously evaluate _pi_. Thus, to check a resulting Q-estimate for _pi_, one can run Monte Carlo episodes under _pi_ and directly compare it.

But how is the situation for Q-Learning? We have some behavior policy _b_ and a target policy _pi_ in off-policy learning. The off-policy evaluation literature you mentioned is, in my opinion, related but different from the topic I am interested in. The papers try to learn about a known _pi_ with limited trajectories under _b_. But do we know the target policy in Q-Learning? In my understanding, the target policy in Q-Learning is not a static one and changes during training. In particular, the current target policy is the greedy policy induced by the current Q-estimates, but since these change over time, the greedy and thus the target policy change.

In my understanding, since we discussed that the Q-estimate in Q-Learning directly approximates Q*, one would have to compare Monte Carlo returns following _pi*_ (the greedy policy induced by Q*, optimal). But we do not know _pi*_ before convergence. How can we still estimate the bias in the course of the training? My intuition would be to get MC returns under the greedy policy of the current Q-estimate, but, as discussed above, this might induce a policy _different_ from the true _pi*_, and we are again comparing apples to pears. It would be great if you could share your thoughts on this issue. I believe my main problem is that I do not fully understand what the target policy in Q-Learning represents.

Best,
Martin

**Comment by watchernyu** [comment]
Thanks again for this very good question! I thought about it a bit more and let me try give another explanation: 

In our case, we define the bias to be Q(s,a) - q^pi(s,a). To avoid confusion, I will use capital Q to indicate a value estimate, and small q for ground truth value. This bias value essentially tells us "how much error is in our current Q estimate at that particular point during training, when the current policy is pi". So here the bias is defined as the current Q estimate value minus the true q value for the current policy (the greedy policy w.r.t. the current Q estimates). At any point in training, when we want to estimate the bias, we stop the learning process, and we take the Q networks and the policy at that point, we keep them fixed, and then run a few MC rollouts using the current policy to get a series of s-a pairs, as well as their current Q estimates and MC returns, during this process these networks are all kept fixed to prevent any changes. In our case, the bias is defined to be related to the current policy (instead of the optimal policy). 

Now in the special case of pi=pi*, then we can treat our Q(s,a) as an estimate of q*(s,a), which is the true q value of the optimal policy. We can again compute Q(s,a) - q*(s,a), which will give us "how much error is in our current Q estimate at that particular point during training, when the current policy is the optimal policy". 

Alternatively we can define the bias in a different way, say, as Q(s,a) - q*(s,a), which will give us "how different is our current Q estimate at that particular point during training, compared to the true Q value of the optimal policy". If we follow this definition, although we don't know the optimal policy beforehand, we can first learn an expert policy, and then use that expert policy to obtain an estimate for the q*. This definition is the one used in the DDQN paper, while the previous definition is used in REDQ and the TD3 paper. 

I guess perhaps the subtle and confusing point here is how we treat Q(s,a): do we treat it as an estimate for the Q value of the current policy (in which case, we should compare to the true q value of the current policy), or as an estimate of the Q value of the optimal policy (in which case, we should compare to the true q value of the optimal policy).

Then there is another question: does it make sense to treat Q(s,a) in these two ways?

Treating Q(s,a) as an estimate of q*(s,a): 1) it makes sense because Q-learning directly approximates q*, so it seems this aligns more with how Q-learning works. 2) however, following this definition, we can only compare the current Q estimate with q* (in the DDQN paper, you will see a horizontal line indicating q*, since q* is fixed). In early stages of training, it is very natural to have large bias values because a random policy gets little positive reward and the Q estimates should be much lower than q*. This definition makes it hard to understand how well the Q estimates are learned in early stages of training. (for example, if we use this definition in SAC or REDQ, we should see a large negative bias in the beginning, which gradually reduces to near 0 at the end of training)

Treating Q(s,a) as an estimate of q^pi(s,a): 1) (seemingly) does not align with how Q-learning works until our policy becomes the optimal policy, in which case this bias definition is the same as the previous definiton. (in fact, although we say "Q-learning approximates q*, we know this will only happen at the end of training when we are near the optimal policy, before that, we only update with limited data and the Q target values are computed with the current greedy policy, which can be far from the optimal policy. So I actually feel this defition makes more sense) 2) the bias values obtained with this method makes a lot more sense, especially during early stages of training. We can observe when agent is stably trained, the current Q values learned are actually very close to the MC returns obtained by the current policy. Such an agent will give a near-zero bias throughout training with this bias definition. 

I hope that helps to some extent, I will also think about this question a bit more :)

**Comment by MarWaltz** [comment]
Great answer, thank you very much! This is getting to the core of my issue. You are right; my actual question is whether to interpret Q(s,a) in Q-Learning as estimating q^pi(s,a) or q*(s,a) in your notation.

But shouldn't we differentiate here between actor-critic methods (SAC, TD3, and modifications) and plain Q-Learning/DQN? For me, in the actor-critic case you mentioned, the algorithms iterate policy evaluation and improvement, e.g., the SAC paper derives the algorithm on this basis. Then it appears clear that the critic indeed estimates q^pi(s,a), where pi is dictated from the actor. I guess this is the reason why in the TD3 and REDQ (which is SAC-based in the respective example) papers, this "Q(s,a) estimates q^pi(s,a)" view is reasonable and perhaps even mandatory. But aren't things a bit different for plain Q-Learning? I know that Q-Learning originates from Value Iteration, which, following Sutton & Barto (2018, p. 83), "combines the policy improvement and truncated policy evaluation steps." So I am wondering whether the q^pi(s,a) estimation view is justifiable in this context. Might this be the reason why the DDQN paper prefers to use the q*(s,a) view?

**Comment by watchernyu** [comment]
No problem! In terms of the difference between actor-critic and plain DQN, my personal feel is that we can perhaps see them as fundamentally the same... for example, DQN only has a Q function (Q network) and does not have a separate policy network. But this is because the policy can be simply inferred from the Q values. If we look at each DQN training step as first updating the Q function, and then setting a policy funtion to map to the actions that maximize the Q function, then this is essentially the same as in actor-critic, where the Q network is updated, and then the actor modifies its proposed actions to be those that maximize the Q. Similarlly, if we look at actor-critic methods, and if we can in fact efficiently compute, for a state s, the Q(s,a) for all action a, then we no longer need a policy (actor); when we need to pick an action in a state s, we can simply look at Q values for all actions, and pick the largest one, then we essentially got plain DQN (with a much larger action space). 

So I would say that the major difference is when we move to continuos action space, we have to modify the implementation a bit to tackle some new challenges, but fundamentally it's still the same Q-learning. So I feel that for this particular issue they are actually the same. 

For DDQN paper, since it is a fairly early paper, and the DQN->DDQN is very simple and effective (they are also the first paper that got very major performance gains in DRL with a bias reduction technique), I'm guessing perhaps the authors found the q*(s,a) view to be sufficient to illustrate their point so they did not try to further analyze it in a more fine-grained way.

**Comment by MarWaltz** [comment]
Okay, it seems like I lack some intuition with Q-Learning since the separation of policy evaluation and improvement makes the whole process very comprehensive for me, but Q-Learning is somehow mixed. 

One final question before I demand too much of your time. You said: "If we look at each DQN training step as first updating the Q function, and then setting a policy function to map to the actions that maximize the Q function,..." So it appears you say that the policy improvement step is deriving a new policy _after the update_. But wouldn't this be SARSA? I thought that, since Q-Learning is off-policy and the behavior policy doesn't matter as long as all (s, a) are visited infinitely often, the policy improvement step in Q-Learning appears _simultaneously_ with the evaluation step during the update. So isn't the improvement step precisely the max-target policy, independent of exploration?

Sorry for asking somewhat basic questions, but I want to understand those relationships fully.

**Comment by watchernyu** [comment]
These are very good questions and discussion! So maybe I was not that clear in that sentence, what I meant to say is:
First, we know that (with some abuse of notations):
1) Q learning update is basically Q(s,a) <- r + gamma max_{a'} Q(s',a')
2) when taking an action in state s in Q learning (let's just say the greedy action), we basically pick the action with highest Q, which is argmax_a Q(s,a)

Now, note that here: max_{a'} Q(s',a')
we are taking the max Q value over all actions of state s'. 

And note that when we select an action, which is argmax_a Q(s,a), we are also selecting an action that maximize the Q value.

So basically, we can replace part of DQN with a policy pi (however, DQN is still DQN, we are only doing it in a slightly different way, the result is still the same)
1) Q learning update is now Q(s,a) <- r + gamma Q(s',pi(s'))
1.5) then we update the policy so that pi(s) = argmax_a Q(s,a) (we can also do this step before previous step)
2) when taking an action in state s, we basically pick pi(s)

So the idea is, we can write DQN in a slightly different way, now the policy becomes explicit instead of implicit, and we will find this new look of DQN is essentially the same (if we ignore the difference in discrete-continuous action spaces) as actor-critic algorithms such as SAC. 

It is a little subtle here... let me know if that makes sense!

**Comment by MarWaltz** [comment]
Excellent! So we can write Q-Learning as:

1. Policy evaluation: Q(s,a) <- r + gamma Q(s', pi(s')) for (s,a)-pair that is updated
2. Policy improvement : pi(s) = argmax_a Q(s,a) for all s

From this perspective, estimating the bias with the q^pi(s, a) view makes more sense since the current Q-estimates evaluate the current greedy policy. Thanks, I will close the topic now. This was a very fruitful discussion!

**Comment by watchernyu** [comment]
A fruitful discussion indeed! :D
